File: packages/silo-client-java/src/test/java/in/org/quicko/silo/client/EntriesTest.java
package in.org.quicko.silo.client;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import in.org.quicko.silo.client.collections.CollectionHandle;
import in.org.quicko.silo.client.entries.Entry;
import in.org.quicko.silo.client.entries.EntryListQuery;
import in.org.quicko.silo.client.entries.EntryPage;
import in.org.quicko.silo.client.entries.EntryReadOptions;
import in.org.quicko.silo.client.query.Filter;
import in.org.quicko.silo.client.query.Sort;
import in.org.quicko.silo.client.support.Post;
import in.org.quicko.silo.client.support.RecordedRequest;
import in.org.quicko.silo.client.support.StubHttp;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class EntriesTest {
  private static final String Row =
      "{\"id\":\"01ABC\",\"rev\":3,\"created_at\":\"2026-09-01T10:00:00.000Z\","
          + "\"updated_at\":\"2026-09-02T11:30:00.000Z\","
          + "\"title\":\"Hello\",\"status\":\"published\",\"tags\":[\"java\"]}";

  private StubHttp server;

  @BeforeEach
  void newStub() {
    server = StubHttp.create();
  }

  private CollectionHandle<Post> posts() {
    return server.silo("k").scope("acme", "prod").collection("posts", Post.class);
  }

  @Test
  void splitsTheFlatRowIntoTheEnvelopeAndTheAuthorsFields() {
    server.enqueueJson(Row);
    Entry<Post> entry = posts().get("01ABC");

    assertEquals("01ABC", entry.id());
    assertEquals(3, entry.rev());
    assertEquals(Instant.parse("2026-09-01T10:00:00Z"), entry.createdAt());
    assertEquals(Instant.parse("2026-09-02T11:30:00Z"), entry.updatedAt());
    assertEquals("Hello", entry.fields().title);
    assertEquals(List.of("java"), entry.fields().tags);
  }

  @Test
  void leavesTheAuthorsOwnKeysExactlyAsTheyArrived() {
    server.enqueueJson(
        "{\"id\":\"01ABC\",\"rev\":1,\"created_at\":\"2026-09-01T10:00:00.000Z\","
            + "\"updated_at\":\"2026-09-01T10:00:00.000Z\",\"product_code\":\"AB-1\"}");

    Entry<Map<String, Object>> entry =
        server.anonymous().scope("acme", "prod").collection("things").get("01ABC");

    assertEquals(Map.of("product_code", "AB-1"), entry.fields());
  }

  @Test
  void resolvesVariablesByDefaultAndAsksForTheTemplatesOnlyWhenTold() {
    server.enqueueJson(Row);
    posts().get("01ABC");
    assertEquals("/api/projects/acme/envs/prod/collections/posts/01ABC", server.lastRequest().target());

    server.enqueueJson(Row);
    posts().get("01ABC", EntryReadOptions.raw());
    assertEquals(
        "/api/projects/acme/envs/prod/collections/posts/01ABC?variables=raw",
        server.lastRequest().target());
  }

  @Test
  void bothWritesAskForTheStoredTemplatesBack() {
    server.enqueueJson(Row);
    posts().create(new Post("Hello", "published", List.of("java")));
    RecordedRequest created = server.lastRequest();
    assertEquals("POST", created.method());
    assertEquals("/api/projects/acme/envs/prod/collections/posts?variables=raw", created.target());
    assertTrue(created.body().contains("\"title\":\"Hello\""));

    server.enqueueJson(Row);
    posts().replace("01ABC", 3, new Post("Hello again", "draft", List.of()));
    RecordedRequest replaced = server.lastRequest();
    assertEquals("PUT", replaced.method());
    assertEquals(
        "/api/projects/acme/envs/prod/collections/posts/01ABC?rev=3&variables=raw",
        replaced.target());
  }

  @Test
  void sendsTheRevisionADeleteIsBoundTo() {
    server.enqueueNoContent();
    posts().delete("01ABC", 7);

    RecordedRequest request = server.lastRequest();
    assertEquals("DELETE", request.method());
    assertEquals("/api/projects/acme/envs/prod/collections/posts/01ABC?rev=7", request.target());
  }

  @Test
  void buildsTheFilterAndSortOntoTheQueryString() {
    server.enqueueJson("{\"data\":[" + Row + "],\"total\":1,\"limit\":50,\"offset\":0}");

    posts().list(EntryListQuery.all()
        .where(Filter.field("status").isEqualTo("published"))
        .sort(Sort.recentlyUpdated())
        .limit(20));

    String target = server.lastRequest().target();
    assertTrue(target.contains("limit=20"), target);
    assertTrue(target.contains("sort=-%24.updated_at"), target);
    assertTrue(target.contains("filter=%7B%22op%22%3A%22eq%22"), target);
  }

  @Test
  void pagesByTheWindowTheServerAnsweredRatherThanTheOneRequested() {
    server.enqueueJson("{\"data\":[" + Row + "],\"total\":600,\"limit\":500,\"offset\":0}");
    EntryPage<Post> page = posts().list(EntryListQuery.all().limit(900));

    assertEquals(500, page.limit());
    assertTrue(page.hasMore());

    server.enqueueJson("{\"data\":[],\"total\":600,\"limit\":500,\"offset\":500}");
    page.next().orElseThrow();

    assertTrue(server.lastRequest().target().contains("offset=500"), server.lastRequest().target());
    assertFalse(server.lastRequest().target().contains("offset=900"));
  }
}

File: packages/silo-client-dart/test/entries_test.dart
import 'package:silo_client/silo_client.dart';
import 'package:test/test.dart';

import 'support/stub_http.dart';

void main() {
  late StubHttp stub;
  late EnvironmentHandle prod;
  setUp(() {
    stub = StubHttp();
    prod = stub.silo().scope('acme', 'prod');
  });

  test('splits the envelope off and renames nothing else', () async {
    stub.json(row('01A', rev: 4, fields: {'product_code': 'X1', 'id_hint': 'kept'}));
    final entry = await prod.collection('posts').get('01A');

    expect(stub.last.url.path, '/api/projects/acme/envs/prod/collections/posts/01A');
    expect(stub.last.url.query, isEmpty);
    expect(entry.id, '01A');
    expect(entry.rev, 4);
    expect(entry.updatedAt, DateTime.utc(2026, 9, 2, 10));
    expect(entry.fields, {'product_code': 'X1', 'id_hint': 'kept'});
  });

  test('reads raw templates when asked', () async {
    stub.json(row('01A'));
    await prod.collection('posts').get('01A', const EntryReadOptions.raw());
    expect(stub.last.url.queryParameters, {'variables': 'raw'});
  });

  test('a converter types the fields both ways, and writes send raw and the revision', () async {
    final posts = prod.collection('posts').withConverter(fromJson: Post.fromJson, toJson: (post) => post.toJson());
    stub
      ..json(row('01A', fields: {'title': 'Hello', 'status': 'draft'}))
      ..json(row('01A', rev: 2, fields: {'title': 'Hello', 'status': 'published'}))
      ..empty();

    final created = await posts.create(const Post('Hello', 'draft'));
    expect(stub.last.method, 'POST');
    expect(stub.last.url.queryParameters, {'variables': 'raw'});
    expect(stub.lastBody, {'title': 'Hello', 'status': 'draft'});

    final replaced = await posts.replace(created.id, created.rev, const Post('Hello', 'published'));
    expect(stub.last.method, 'PUT');
    expect(stub.last.url.query, 'rev=1&variables=raw');
    expect(replaced.fields.status, 'published');

    await posts.delete(replaced.id, replaced.rev);
    expect(stub.last.method, 'DELETE');
    expect(stub.last.url.query, 'rev=2');
  });

  test('sends the filter as JSON and the sort as its string, in the order set', () async {
    stub.json(page([]));
    await prod
        .collection('posts')
        .list(
          EntryListQuery(where: Filter.field('status').equals('published'), sort: Sort.recentlyUpdated(), limit: 20),
        );

    expect(
      stub.last.url.query,
      'limit=20&filter=${Uri.encodeComponent('{"op":"eq","path":"\$.data.status","value":"published"}')}'
      '&sort=-%24.updated_at',
    );
  });

  test('navigates by the window the server answered, not the one asked for', () async {
    stub
      ..json(page([row('01A')], total: 1200, limit: 500))
      ..json(page([row('01B')], total: 1200, limit: 500, offset: 500));

    final first = await prod.collection('posts').list(const EntryListQuery(limit: 900));
    expect(first.pageCount, 3);
    expect(first.hasMore, isTrue);

    final second = await first.next();
    expect(stub.last.url.queryParameters, {'limit': '500', 'offset': '500'});
    expect(second!.entries.single.id, '01B');
    expect(second.pageNumber, 2);
  });

  test('all() pages lazily and stops on the first short page', () async {
    stub
      ..json(page([row('01A'), row('01B')], total: 3, limit: 2))
      ..json(page([row('01C')], total: 3, limit: 2, offset: 2));

    final ids = await prod.collection('posts').all(const EntryListQuery(limit: 2)).map((entry) => entry.id).toList();

    expect(ids, ['01A', '01B', '01C']);
    expect(stub.requests, hasLength(2));
  });

  test('pages() yields whole pages', () async {
    stub
      ..json(page([row('01A'), row('01B')], total: 3, limit: 2))
      ..json(page([row('01C')], total: 3, limit: 2, offset: 2));

    final sizes = await prod
        .collection('posts')
        .pages(const EntryListQuery(limit: 2))
        .map((page) => page.length)
        .toList();
    expect(sizes, [2, 1]);
  });

  test('a cancelled stream stops before the next page', () async {
    final signal = CancellationSignal()..cancel();
    await expectLater(
      prod.collection('posts').all(const EntryListQuery(), EntryReadOptions(cancellation: signal)).toList(),
      throwsA(isA<RequestAbortedException>()),
    );
    expect(stub.requests, isEmpty);
  });
}
